<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />

<meta name="author" content="Hong Zhang and Kai Zhang" />


<title>Introduction to CCMO</title>

<style type="text/css">
a.anchor-section {margin-left: 10px; visibility: hidden; color: inherit;}
a.anchor-section::before {content: '#';}
.hasAnchor:hover a.anchor-section {visibility: visible;}
</style>
<script>// Anchor sections v1.0 written by Atsushi Yasumoto on Oct 3rd, 2020.
document.addEventListener('DOMContentLoaded', function() {
  // Do nothing if AnchorJS is used
  if (typeof window.anchors === 'object' && anchors.hasOwnProperty('hasAnchorJSLink')) {
    return;
  }

  const h = document.querySelectorAll('h1, h2, h3, h4, h5, h6');

  // Do nothing if sections are already anchored
  if (Array.from(h).some(x => x.classList.contains('hasAnchor'))) {
    return null;
  }

  // Use section id when pandoc runs with --section-divs
  const section_id = function(x) {
    return ((x.classList.contains('section') || (x.tagName === 'SECTION'))
            ? x.id : '');
  };

  // Add anchors
  h.forEach(function(x) {
    const id = x.id || section_id(x.parentElement);
    if (id === '') {
      return null;
    }
    let anchor = document.createElement('a');
    anchor.href = '#' + id;
    anchor.classList = ['anchor-section'];
    x.classList.add('hasAnchor');
    x.appendChild(anchor);
  });
});
</script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>


<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(data-line-number);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>



<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">Introduction to CCMO</h1>
<h4 class="author">Hong Zhang and Kai Zhang</h4>
<h4 class="date">2021/02/24</h4>



<div id="methods" class="section level1">
<h1>1. Methods</h1>
<p>The R package <strong>CCMO</strong> implements two powerful statistical methods for disease-gene association analysis using case-control mother-offspring data.</p>
<div id="data-description" class="section level2">
<h2>1.2 Data description</h2>
<p>The data for each mother-offspring pair are <span class="math inline">\((D,G^c,G^m,X)\)</span>, where <span class="math inline">\(D\)</span> is the disease status of the offspring (should be coded as 1 for case and 0 for control), <span class="math inline">\(G^c\)</span> and <span class="math inline">\(G^m\)</span> are the genotypes of offspring and mother, and <span class="math inline">\(X\)</span> is a <span class="math inline">\(p\)</span>-vector of maternal covariates.</p>
Suppose a SNP has two alleles A and B (B is the minor allele), then genotypes can be coded either as 0 (for genotype AA), 1 (for genotype AB), or 2 (for genotype BB). Alternatively, the genotypes can be normalized according to specified mode of inheritance

<p>In what follows, we adopt the normalized genotypes.</p>
</div>
<div id="testing-maternal-and-offspring-genetic-effects-and-envirionment-effects" class="section level2">
<h2>1.3 Testing maternal and offspring genetic effects and envirionment effects</h2>
<div id="reference" class="section level3">
<h3>1.3.1 Reference</h3>
<p>Zhang H, Mukherjee B, Arthur V, Hu G, Hochner H, Chen J (2020). An Efficient and Computationally Robust Statistical Method for Analyzing Case-Control Mother-Offspring Pair Genetic Association Studies. <em>Annals of Applied Statistics</em> <strong>14</strong>: 560–-584.</p>
</div>
<div id="model-description" class="section level3">
<h3>1.3.2 Model description</h3>
<p>The method assumes that the offspring genotype is conditionally independent of the maternal covariates given maternal covariates. The penetrance model is as follows: <span class="math display">\[\begin{eqnarray*}
&amp;&amp;\mbox{pr}(D=1|G^o,G^m,X)\\
&amp;=&amp;\mbox{expit}(\beta_0+\beta_{G^c}G^c+\beta_{G^m}G^m+\beta_oX_o+\beta_{G^cX}G^cX_c+\beta_{G^mX}G^mX_m+\beta_{G^cG^m}G^cG^m),
\end{eqnarray*}\]</span> where expit<span class="math inline">\((t)=e^t/(1+e^t)\)</span>, <span class="math inline">\(X_o\)</span>, <span class="math inline">\(X_c\)</span>, and <span class="math inline">\(X_m\)</span> are functions of the covariates <span class="math inline">\(X\)</span>, <span class="math inline">\(\beta_{G^c}\)</span> and <span class="math inline">\(\beta_{G^m}\)</span> are main genetic effects for offspring (child) and mother, respectively, <span class="math inline">\(\beta_o\)</span> is the main maternal environmental effect, <span class="math inline">\(\beta_{G^cX}\)</span> is the offspring gene-environment interaction effct, <span class="math inline">\(\beta_{G^mX}\)</span> is the maternal gene-environment interaction effect, and <span class="math inline">\(\beta_{G^cG^m}\)</span> is the gene-gene interaction effect. The maternal genotype is associated with the maternal covariates through the following double additive logit (daLOG) model: <span class="math display">\[\mbox{pr}(G^m=1|X)=\frac{\xi(\theta,F)e^{k\eta'X_{G^m}}}{\sum_l\xi_l(\theta,F)e^{l\eta'X_{G^m}}},\]</span> where <span class="math inline">\(\eta\)</span> is a regression parameter vector, <span class="math inline">\(X_{G^m}\)</span> is a function of <span class="math inline">\(X\)</span>, <span class="math inline">\(\xi_k(\theta,F)\)</span> is the probability of <span class="math inline">\(G^m=k\)</span> with <span class="math inline">\(\theta\)</span> being the minor allele frequency and <span class="math inline">\(F\)</span> being the inbreeding coefficient.</p>
</div>
<div id="usage-of-main-functions-singlesnp-and-multiplesnp" class="section level3">
<h3>1.3.4 Usage of main functions <em>singleSNP</em> and <em>multipleSNP</em></h3>

<p>Users should specify <span class="math inline">\(X_c\)</span>, <span class="math inline">\(X_m\)</span>, <span class="math inline">\(X_o\)</span>, <span class="math inline">\(X_{G^m}\)</span> according to their requirements, each of them can be omitted in the model by setting it to be the default value <em>NULL</em>. Each of the two main genetic effects can be included in the penetrance model by specifying <em>G.main</em>, and the gene-gene interaction effect can be included by setting <em>G.int</em> to be <em>TRUE</em>.</p>
<p>The mode of inheritance can be specified through <em>mode</em> (‘rec’ for recessive, ‘add’ for additive, ‘dom’ for dominant). Hardy-Weiberg equilibrium can be incorporated in the method through <em>HWE</em>. The independence of <span class="math inline">\(G^m\)</span> and <span class="math inline">\(X\)</span> can be incorprated by setting <em>ind</em> to be <em>TRUE</em>.</p>
</div>
</div>
<div id="testing-parent-of-origin-effects" class="section level2">
<h2>1.4 Testing parent-of-origin effects</h2>
<div id="reference-1" class="section level3">
<h3>1.4.1 Reference</h3>
<p>Zhang K, Zhang H, Chen J. (2021). Covariate Adjusted Inference of Parent-of-Origin Effects Using Case-Control Mother-Child Paired Multi-Locus Genotype Data. Manuscript.</p>
</div>
<div id="model-description-1" class="section level3">
<h3>1.4.2 Model description</h3>
<p>The penetrance model for parent-of-origin effects (POEs) is as follows: <span class="math display">\[\begin{eqnarray*}
&amp;&amp;\mbox{pr}(D=1|G^c,G^m,G^c_m,G^c_p,X)\\
&amp;=&amp;\mbox{expit}(\beta_0+\beta_{G^c}G^c+\beta_{G^m}G^m+\beta_{POE}(G^c_m-G^c_p)+\beta_{X}X),
\end{eqnarray*}\]</span> <span class="math inline">\(\beta_{G^c}\)</span> and <span class="math inline">\(\beta_{G^m}\)</span> are main genetic effects for offspring (child) and mother, respectively, <span class="math inline">\(\beta_X\)</span> is the main maternal environmental effect, <span class="math inline">\(\beta_{POE}\)</span> is the POE. The genotypes should be coded either as 0 (for genotype AA), 1 (for genotype AB), or 2 (for genotype BB) and we consider using haplotypes data to improve the inference efficiency for assessing POEs. In additional to multi-locus genotypes, disease statuses, and covariates, the inputs for the POE analysis function <em>MultiLociPOE</em> shoud include possible haplotypes in the population of interest, which can be generated by the function <em>MultiLociPOE.input</em>.</p>
</div>
<div id="usage-of-main-function-multilocipoe" class="section level3">
<h3>1.4.3 Usage of main function <em>MultiLociPOE</em></h3>
<p>The usage of the function <em>MultiLociPOE</em> is as follows:</p>

<p>Users should specify test loci and prevalence , and make sure that the genotypes are coded as 0 (for genotype AA), 1 (for genotype AB), or 2 (for genotype BB).</p>

<p>Users can use <em>MultiLociPOE.input</em> to generate inputs for <em>MultiLociPOE</em>. Users should specify <em>type</em> corresponding to the format of the genotype data. Here, ‘0’ means the genotypes are coded as 0, 1, 2; ‘1’ means the genotypes are normalized as 0, 1/2, 1; ‘2’ means the genotypes are coded as something like “A/A”, “A/B”, “B/B”.</p>
</div>
</div>
</div>
<div id="illustration-of-main-functions" class="section level1">
<h1>2. Illustration of main functions</h1>
<div id="singlesnp" class="section level2">
<h2>2.1 <em>singleSNP</em></h2>
<p>We use the dataset contained in the R package for illustration.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="kw">library</span>(CCMO)</a>
<a class="sourceLine" id="cb1-2" data-line-number="2"><span class="kw">data</span>(<span class="st">&quot;SampleData&quot;</span>,<span class="dt">package=</span><span class="st">&quot;CCMO&quot;</span>)</a>
<a class="sourceLine" id="cb1-3" data-line-number="3"><span class="kw">dim</span>(SampleData)</a>
<a class="sourceLine" id="cb1-4" data-line-number="4"><span class="kw">head</span>(SampleData)</a></code></pre></div>
<p>This dataset has maternal and offspring genotypes for 10 SNPs and two maternal covariates from 2000 mother-offspring pairs. Here we analyze the first SNP. Suppose we are interested in two main genetic effects, maternal gene-environment interaction effects, and offspring gene-environment interaction effects. The R code for analyzing the data is as follows:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" data-line-number="1"><span class="co"># library(CCMO)</span></a>
<a class="sourceLine" id="cb2-2" data-line-number="2">Y =<span class="st"> </span>SampleData[,<span class="dv">1</span>]</a>
<a class="sourceLine" id="cb2-3" data-line-number="3">Gc =<span class="st"> </span>SampleData[,<span class="dv">2</span>]</a>
<a class="sourceLine" id="cb2-4" data-line-number="4">Gm =<span class="st"> </span>SampleData[,<span class="dv">12</span>]</a>
<a class="sourceLine" id="cb2-5" data-line-number="5">X =<span class="st"> </span>SampleData[,<span class="op">-</span>(<span class="dv">1</span><span class="op">:</span><span class="dv">21</span>)]</a>
<a class="sourceLine" id="cb2-6" data-line-number="6">fit =<span class="st"> </span><span class="kw">singleSNP</span>(Y,Gc,Gm,<span class="dt">Xo=</span>X,<span class="dt">Xc=</span>X,<span class="dt">Xm=</span>X,<span class="dt">X.Gm=</span>X)</a></code></pre></div>
<p>The result is a list, whose elements can be accessed through</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" data-line-number="1"><span class="kw">names</span>(fit)</a></code></pre></div>
<p>The estimation and significance test results produced by the new method are stored in</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" data-line-number="1">fit<span class="op">$</span>new</a></code></pre></div>
<p>The covariance matrix of the estimates by the new method is stored in</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" data-line-number="1">fit<span class="op">$</span>cov.new</a></code></pre></div>
<p>The corresponding results by the standard logistic regression are included for comparison</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" data-line-number="1">fit<span class="op">$</span>log</a>
<a class="sourceLine" id="cb6-2" data-line-number="2">fit<span class="op">$</span>cov.log</a></code></pre></div>
</div>
<div id="omnibustest" class="section level2">
<h2>2.2 <em>OmnibusTest</em></h2>
<p>Multiple effects can be simultineously tested using the function <em>OmnibusTest</em>. Suppose we want to test four gene-environment interaction effects simultineously, then we should specify the indices of the two interaction effects when calling <em>OmnibusTest</em>:</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" data-line-number="1">fit =<span class="st"> </span><span class="kw">OmnibusTest</span>(fit,<span class="dt">test=</span><span class="dv">7</span><span class="op">:</span><span class="dv">10</span>)</a>
<a class="sourceLine" id="cb7-2" data-line-number="2">fit<span class="op">$</span>Omnibus</a></code></pre></div>
<p>The results include the Wald test statistic, degrees of freedom, and p-value for each of two methods (i.e., <em>new</em> and <em>log</em>).</p>
</div>
<div id="multiplesnp" class="section level2">
<h2>2.3 <em>multipleSNP</em></h2>
<p>Suppose we want to analyze the 10 SNPs simultaneously. The function <em>multipleSNP</em> calls multiple CPU cores to save computational time, as shown below</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb8-1" data-line-number="1">Gc =<span class="st"> </span>SampleData[,<span class="dv">2</span><span class="op">:</span><span class="dv">11</span>]</a>
<a class="sourceLine" id="cb8-2" data-line-number="2">Gm =<span class="st"> </span>SampleData[,<span class="dv">12</span><span class="op">:</span><span class="dv">21</span>]</a>
<a class="sourceLine" id="cb8-3" data-line-number="3"><span class="kw">system.time</span>(fit1 &lt;-<span class="st"> </span><span class="kw">multipleSNP</span>(Y,Gc,Gm,<span class="dt">Xo=</span>X,<span class="dt">Xc=</span>X,<span class="dt">Xm=</span>X,<span class="dt">X.Gm=</span>X,<span class="dt">cores=</span><span class="dv">1</span>))</a>
<a class="sourceLine" id="cb8-4" data-line-number="4"><span class="kw">system.time</span>(fit2 &lt;-<span class="st"> </span><span class="kw">multipleSNP</span>(Y,Gc,Gm,<span class="dt">Xo=</span>X,<span class="dt">Xc=</span>X,<span class="dt">Xm=</span>X,<span class="dt">X.Gm=</span>X,<span class="dt">cores=</span><span class="dv">2</span>))</a></code></pre></div>
<p>Of course, the saved computational time depends on the number of CPU cores. The output is a list of length 10 (the number of SNPs). For example, the results of the 2nd SNP can be accessed through</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb9-1" data-line-number="1">fit2[[<span class="dv">2</span>]]</a></code></pre></div>
<p>Omnibus test can be carried out by specifying the target variable indices through <em>test</em>:</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb10-1" data-line-number="1">fit &lt;-<span class="st"> </span><span class="kw">multipleSNP</span>(Y,Gc,Gm,<span class="dt">Xo=</span>X,<span class="dt">Xc=</span>X,<span class="dt">Xm=</span>X,<span class="dt">X.Gm=</span>X,<span class="dt">test=</span><span class="dv">7</span><span class="op">:</span><span class="dv">10</span>)</a>
<a class="sourceLine" id="cb10-2" data-line-number="2">fit[[<span class="dv">2</span>]]<span class="op">$</span>Omnibus</a></code></pre></div>
</div>
<div id="multilocipoe" class="section level2">
<h2>2.4 <em>MultiLociPOE</em></h2>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb11-1" data-line-number="1"><span class="co"># library(CCMO)</span></a>
<a class="sourceLine" id="cb11-2" data-line-number="2">Y =<span class="st"> </span>POESampleData[,<span class="dv">1</span>]</a>
<a class="sourceLine" id="cb11-3" data-line-number="3">gmm =<span class="st"> </span>POESampleData[,<span class="dv">2</span><span class="op">:</span><span class="dv">6</span>]</a>
<a class="sourceLine" id="cb11-4" data-line-number="4">gcc =<span class="st"> </span>POESampleData[,<span class="dv">7</span><span class="op">:</span><span class="dv">11</span>]</a>
<a class="sourceLine" id="cb11-5" data-line-number="5">X =<span class="st"> </span>POESampleData[,<span class="dv">12</span>]</a>
<a class="sourceLine" id="cb11-6" data-line-number="6">data =<span class="st"> </span><span class="kw">MultiLociPOE.input</span>(gmm,gcc,<span class="dv">0</span>)</a>
<a class="sourceLine" id="cb11-7" data-line-number="7">gmm =<span class="st"> </span>data<span class="op">$</span>gmm</a>
<a class="sourceLine" id="cb11-8" data-line-number="8">gcc =<span class="st"> </span>data<span class="op">$</span>gcc</a>
<a class="sourceLine" id="cb11-9" data-line-number="9">hap =<span class="st"> </span>data<span class="op">$</span>hap</a>
<a class="sourceLine" id="cb11-10" data-line-number="10">ppi =<span class="st"> </span>data<span class="op">$</span>ppi</a>
<a class="sourceLine" id="cb11-11" data-line-number="11">loci =<span class="st"> </span><span class="dv">1</span></a>
<a class="sourceLine" id="cb11-12" data-line-number="12">f =<span class="st"> </span><span class="fl">0.01</span></a>
<a class="sourceLine" id="cb11-13" data-line-number="13">fit =<span class="st"> </span><span class="kw">MultiLociPOE</span>(Y,gmm,gcc,X,loci,hap,f,ppi)</a></code></pre></div>
<p>The result is a list, whose elements can be accessed through</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb12-1" data-line-number="1"><span class="kw">names</span>(fit)</a></code></pre></div>
<p>The estimation and significance test results produced by the new method are stored in</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb13-1" data-line-number="1">fit<span class="op">$</span>new</a></code></pre></div>
<p>The covariance matrix of the estimates by the new method is stored in</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb14-1" data-line-number="1">fit<span class="op">$</span>cov.new</a></code></pre></div>
<p>The corresponding results by the standard logistic regression are included for comparison</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb15-1" data-line-number="1">fit<span class="op">$</span>log</a>
<a class="sourceLine" id="cb15-2" data-line-number="2">fit<span class="op">$</span>cov.log</a></code></pre></div>
</div>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
